### PR TITLE
Features aug 2019

### PR DESCRIPTION
**Pull request by nkunal**
Add Histogram support
New Transforms
Several bug fixes
Upgrade kafka to 2.x
Upgrade Elasticsearch to 6.x
Add retention feature for schema docs 
Migrate annotation to elasticsearch
Add pagerduty notifier
Increase code coverage
Enhance datalag detection and disabling of alerts
Add Image chart feature to alert notifications

### CODE REVIEW

**Comment by nkunal** [comment]
run it

**Comment by nkunal** [comment]
@dilipdevaraj-sfdc please review. One UT is occasionally failing. Probably not an issue with the code but the test itself

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/MonitorService.java [hunk 1]
```diff
@@ -190,71 +189,150 @@
...
+
+        CONSUMER_OFFSET_RECORDS_WRITE_FAILURES("argus.core", "consumer.offset.records.write.failures", MetricType.COUNTER),
+        CONSUMER_OFFSET_RECORDS_WRITE_LATENCY("argus.core", "consumer.offset.records.write.latency", MetricType.COUNTER),
+        CONSUMER_OFFSET_RECORDS_READ_LATENCY("argus.core", "consumer.offset.records.read.latency", MetricType.COUNTER),
 
-        BLOOMFILTER_APPROXIMATE_ELEMENT_COUNT("argus.core", "bloomfilter.approximate.element.count"),
-        BLOOMFILTER_SCOPE_ONLY_APPROXIMATE_ELEMENT_COUNT("argus.core", "bloomfilter.scope.only.approximate.element.count"),
-        BLOOMFILTER_SCOPE_AND_METRIC_ONLY_APPROXIMATE_ELEMENT_COUNT("argus.core", "bloomfilter.scope.and.metric.only.approximate.element.count"),
-        BLOOMFILTER_METATAGS_APPROXIMATE_ELEMENT_COUNT("argus.core", "bloomfilter.metatags.approximate.element.count");
+        IDB_CLIENT_GET_LATENCY("argus.core", "idb.client.get.latency", MetricType.COUNTER),
```
Comment (colbyguan) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/MonitorService.java]: These can be removed too

File: pom.xml [hunk 2]
```diff
@@ -103,31 +103,34 @@
...
         <junit.version>4.12</junit.version>
+        <jacoco-maven-plugin.version>0.7.9</jacoco-maven-plugin.version>
+        <jmx.prometheus.version>0.11.0</jmx.prometheus.version>
     </properties>
+     <!-- Add repositories here instead of maintaining in maven-settings.xml
+         Please add <server> information in ~/.m2/settings.xml -->
+    <repositories>
+        <repository>
+            <id>nexus</id>
+            <url>https://nexus.soma.salesforce.com/nexus/content/groups/public</url>
```
Comment (colbyguan) [inline pom.xml]: Removeable?

File: pom.xml [hunk 3]
```diff
@@ -304,69 +334,12 @@
...
-                            </execution>
-                        </executions>
-                    </plugin>
-                </plugins>
-            </build>
-        </profile>
-    </profiles>
+
+    <scm>
+      <connection>scm:git:ssh://git@git.soma.salesforce.com/ArgusMonitoring/Argus</connection>
```
Comment (colbyguan) [inline pom.xml]: Removeable block?

File: pom.xml [hunk 4]
```diff
@@ -4,13 +4,13 @@
...
     <artifactId>argus</artifactId>
-    <version>2.22.3</version>
+    <version>4-SNAPSHOT</version>
     <packaging>pom</packaging>
 
     <name>Argus</name>
     <description>A pom for all Argus projects. It provides standards for plugin versions, configuration and a number of other things.</description>
     <inceptionYear>2016</inceptionYear>
-    <url>https://www.github.com/salesforce/argus</url>
+    <url>https://git.soma.salesforce.com/ArgusMonitoring/Argus</url>
```
Comment (dilipdevaraj-sfdc) [inline pom.xml]: should we specifying git.soma.salesforce.com in open source ?

File: pom.xml [hunk 5]
```diff
@@ -424,21 +397,20 @@
...
     <distributionManagement>
-        <snapshotRepository>
-            <id>ossrh</id>
-            <url>https://oss.sonatype.org/content/repositories/snapshots</url>
-        </snapshotRepository>
         <repository>
-            <id>ossrh</id>
-            <url>https://oss.sonatype.org/service/local/staging/deploy/maven2/</url>
+            <id>nexus</id>
+            <url>https://nexus.soma.salesforce.com/nexus/content/repositories/releases/</url>
```
Comment (dilipdevaraj-sfdc) [inline pom.xml]: removable ?

File: pom.xml [hunk 6]
```diff
@@ -424,21 +397,20 @@
     </dependencyManagement>
     <modules>
         <module>ArgusCore</module>
-        <module>ArgusWeb</module>
         <module>ArgusWebServices</module>
         <module>ArgusClient</module>
         <module>ArgusSDK</module>
+        <!--<module>ArgusQuerySnapshots</module>-->
+        <!--ArgusQuerySnaphots is a child directory that does not depend on the latest ArgusCore, not needs to be built frequently-->
```
Comment (dilipdevaraj-sfdc) [inline pom.xml]: can remove comment on ArgusQuerySnaphots

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/MonitorService.java (hunk 1 above)
Comment (nkunal) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/MonitorService.java]: removed

File: pom.xml (hunk 2 above)
Comment (nkunal) [reply pom.xml]: removed

File: pom.xml (hunk 3 above)
Comment (nkunal) [reply pom.xml]: removed

File: pom.xml (hunk 4 above)
Comment (nkunal) [reply pom.xml]: removed

File: pom.xml (hunk 5 above)
Comment (nkunal) [reply pom.xml]: removed

File: pom.xml (hunk 6 above)
Comment (nkunal) [reply pom.xml]: removed

**Comment by nkunal** [comment]
@dilipdevaraj-sfdc @colbyguan all comments addressed, please approve

**Comment by colbyguan** [comment]
@nkunal  There are a set of `*.sfdc.net` hostnames in unit test mocks if we care about those 
Additionally it looks like some of the Refocus related stuff eg. Forwarder are still there

**Comment by dilipdevaraj-sfdc** [comment]
@nkunal 
Can you please remove these references

ArgusCore/target/test-classes/com/salesforce/dva/argus/service/metric/MetricReaderTest.testFunctionsWithConstant.properties:func-with-regex-constant=AVERAGEBELOW(-1d:na1:app_record.count:avg, #(app1-1\\.ops\\.sfdc\\.net)#)
ArgusCore/src/test/resources/com/salesforce/dva/argus/service/metric/MetricReaderTest.testFunctionsWithConstant.properties:func-with-regex-constant=AVERAGEBELOW(-1d:na1:app_record.count:avg, #(app1-1\\.ops\\.sfdc\\.net)#)
ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/ElasticSearchConsumerOffsetMetricsServiceTest.java:			"          \"key\" : \"sfdc.prod.ajna__prd.ajna_local__metrics\",\n" +
ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/ElasticSearchConsumerOffsetMetricsServiceTest.java:			"          \"key\" : \"sfdc.prod.ajna__phx.ajna_local__metrics\",\n" +
ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/ElasticSearchConsumerOffsetMetricsServiceTest.java:				"[namespace=>null, scope=>ajna.consumer, metric=>metric.consumer.lag, tags=>{topic=sfdc.prod.ajna__phx.ajna_local__metrics}, datapoints=>{1560902400000=21455.0, 1561075200000=21981.0, 1561248000000=22461.0}, " +
ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/ElasticSearchConsumerOffsetMetricsServiceTest.java:				"namespace=>null, scope=>ajna.consumer, metric=>metric.consumer.lag, tags=>{topic=sfdc.prod.ajna__prd.ajna_local__metrics}, datapoints=>{1560902400000=19676.0, 1561075200000=21635.0, 1561248000000=22659.0}]";
ArgusCore/src/main/java/com/salesforce/dva/argus/system/SystemConfiguration.java:        PKI_MONITORED_DIRECTORY("pki.monitored.directory", "/etc/pki_service/sfdc/argus-client"),

**Comment by dilipdevaraj-sfdc** [comment]
@nkunal 
pls remove "prd" references

ArgusCore/src/test/java/com/salesforce/dva/argus/util/AlertUtilsTest.java:				"DOWNSAMPLE(CULL_BELOW(DERIVATIVE(-115m:-15m:iot-container.PRD.NONE.-:iot.flows.state.load.errors_count{flowsnakeEnvironmentName=iot-prd-stmfa-00ds70000000mqy}:zimsum:1m-sum), #0#, #value#), #10m-sum#)",
ArgusCore/src/test/java/com/salesforce/dva/argus/util/AlertUtilsTest.java:				"DOWNSAMPLE(CULL_BELOW(DERIVATIVE(1418312700000:1418318700000:iot-container.PRD.NONE.-:iot.flows.state.load.errors_count{flowsnakeEnvironmentName=iot-prd-stmfa-00ds70000000mqy}:zimsum:1m-sum),#0#,#value#),#10m-sum#)",
ArgusCore/src/test/java/com/salesforce/dva/argus/service/RefocusServiceTest.java:    private static final String refocus_prd_proxy = "myhostname.abc.com";
ArgusCore/src/test/java/com/salesforce/dva/argus/service/RefocusServiceTest.java:    private static final String refocus_prd_proxy_port = "8080";
ArgusCore/src/test/java/com/salesforce/dva/argus/service/RefocusServiceTest.java:                _configuration.setProperty(RefocusProperty.REFOCUS_PROXY_HOST.getName(), refocus_prd_proxy); // e.g. myhostname.abc.com
ArgusCore/src/test/java/com/salesforce/dva/argus/service/RefocusServiceTest.java:                _configuration.setProperty(RefocusProperty.REFOCUS_PROXY_PORT.getName(), refocus_prd_proxy_port);
ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/ElasticSearchConsumerOffsetMetricsServiceTest.java:			"          \"key\" : \"sfdc.prod.ajna__prd.ajna_local__metrics\",\n" +
ArgusCore/src/test/java/com/salesforce/dva/argus/service/metric/ElasticSearchConsumerOffsetMetricsServiceTest.java:				"namespace=>null, scope=>ajna.consumer, metric=>metric.consumer.lag, tags=>{topic=sfdc.prod.ajna__prd.ajna_local__metrics}, datapoints=>{1560902400000=19676.0, 1561075200000=21635.0, 1561248000000=22659.0}]";

**Comment by dilipdevaraj-sfdc** [comment]
@nkunal 
pls remove references to "gus.my"

ArgusCore/src/test/java/com/salesforce/dva/argus/service/alert/notifier/GusNotifierIT.java:        p.setProperty("notifier.property.alert.gus_endpoint", "https://gus.my.salesforce.com/services/oauth2/token");
ArgusCore/src/test/java/com/salesforce/dva/argus/service/alert/notifier/GusNotifierIT.java:        p.setProperty("notifier.property.alert.gus_post_endpoint", "https://gus.my.salesforce.com/services/data/v35.0/chatter/feed-elements?feedElementType=FeedItem");
ArgusCore/src/test/java/com/salesforce/dva/argus/service/alert/notifier/GusTransportIT.java:    private String AUTH_ENDPOINT = "https://gus.my.salesforce.com/services/oauth2/token";
ArgusCore/src/test/java/com/salesforce/dva/argus/service/alert/notifier/GusTransportIT.java:        assertEquals("https://gus.my.salesforce.com", ei.getEndPoint());

**Comment by dilipdevaraj-sfdc** [comment]
@nkunal 
In testFunctionsWithConstant.properties https://github.com/salesforce/Argus/pull/1060/commits/e212f61f0299dfbcefaaf3786a70d5fb85a32cd2#diff-7b155a4727b5702536fa86ffe88c5f91   there are DC information.
SFZ.AGG.none
AJNA-CHI-SP1-SFZ
We need to remove that. Please search if they are present in some other files too.

For reference the original open source file was
https://github.com/salesforce/Argus/blob/develop/ArgusCore/src/test/resources/com/salesforce/dva/argus/service/metric/MetricReaderTest.testFunctionsWithConstant.properties

**Pull request merged**